floodlight/io/catapultApi.py

```diff
@@ -0,0 +1,384 @@
+            raise APIRequestError(f"API request failed with status code {response_data_json_i.status_code}")
+
+    return response_data_json_list
+
+def get_key_names_from_json(response_data_json_list: List[Dict]) -> set:
+    """
+    Extracts unique column names from a list of JSON data responses.
+
+    Args:
+        response_data_json_list (List[Dict]): List of JSON data responses.
+
+    Returns:
+        set: A set of unique column names.
+    """
+    recorded_columns = set()
+
+    for data_i in response_data_json_list:
+        recorded_columns.update(data_i.keys())
+        if "data" in data_i and isinstance(data_i["data"], list) and data_i["data"]:
+            recorded_columns.update(data_i["data"][0].keys())
+
+    return recorded_columns
+
+def _get_key_links(response_data_json_list: List[Dict]) -> Union[Dict[str, int], None]:
+    """
+    Generates a mapping of key names to their indices in the JSON data responses.
+
+    Args:
+        response_data_json_list (List[Dict]): List of JSON data responses.
+
+    Returns:
+        Union[Dict[str, int], None]: A dictionary mapping key names to their indices or None if critical keys are missing.
+    """
+    for data_i in response_data_json_list:
+        data_i['name'] = f"{data_i['athlete_first_name']} {data_i['athlete_last_name']}"
+
+    key_names = list(get_key_names_from_json(response_data_json_list))
+
+    mapping = {
+        "ts": "time",
+        "device id": "sensor_id",
+        "athlete_id": "mapped_id",
+        "athlete_first_name": "athlete_first_name",
+        "athlete_last_name": "athlete_last_name",
+        "stream_type": "stream_type",
+        "name": "name",
+        "jersey": "number",
+        "team_id": "group_id",
+        "team_name": "group_name",
+        "x": "x_coord",
+        "y": "y_coord",
+        "cs": "cs",
+        "lat": "lat",
+        "long": "long",
+        "o": "o",
+        "v": "v",
+        "a": "a",
+        "hr": "hr",
+        "pl": "pl",
+        "mp": "mp",
+        "sl": "sl",
+    }
+
+    necessary_keys = ["time", "x_coord", "y_coord"]
+    key_links = {}
+    for key in mapping:
+        if key in key_names:
+            key_links.update({mapping[key]: key_names.index(key)})
+
+    if not all(columns in key_links for columns in necessary_keys):
+        warnings.warn("Data file lacks critical information! No timestamp or coordinates found.")
+        return None
+
+    return key_links
+
+def get_meta_data(response_data_json_list: List[Dict]) -> Tuple[Dict[str, Dict[str, List[str]]], int, int, int]:
+    """
+    Extracts metadata from the JSON data responses.
+
+    Args:
+        response_data_json_list (List[Dict]): List of JSON data responses.
+
+    Returns:
+        Tuple[Dict[str, Dict[str, List[str]]], int, int, int]: 
+            - Dictionary mapping group IDs to player identifiers.
+            - Number of frames.
+            - Framerate.
+            - Null timestamp.
+    """
+    framerate = 10

```

Comment:
This is not very elegant. Altough we can assume the framerate is 10 Hz. This may not be always true in the future. Please implement a basic framerate detection as you can see in the `get_meta_data()` function in the Kinexon parser.

Reply:
Done